### DISCUSSION TITLE
normalize intensity

### DISCUSSION THREAD

**Discussion by wiesehahn** [r-lidar/lasR#87] [General]
Just a thought:
The intensity measurement seems quite underused in lidar applications to me. This is probably due to the fact that the values are not as trivial as XYZ. They are dependent on multiple factors in scene. And additionally their scale varies by sensor. Even unnormalized values showed value to me, but there is even more potential if they would be normalized.

I am wondering if it would be possible and make sense to add a stage to normalize intensities? Probably similar to the implementation in lidR. 

Additionally to this range dependent normalization I am wondering if it would be a good idea to further scale/normalize intensities to some range (e.g. 0-1) based on something like min and max or percentile intensity values per file? This would possibly allow not just for better comparison within a file but also between campaigns.

**Comment by Jean-Romain** [comment]
The goal of `lasR` is not to allow R&D like `lidR` but rather to provide efficient implementation of methods that proved to work well. In `lidR` I'm keen to add stuff that may not work not that well so people can try and explore in a R&D fashion.

Sensor tracking in `lidR` is highly dependent on the quality of the LAS file. It works well but you can find numerous issues on github related to that algorithm that were caused by LAS file not following the specifications and having weird duplicated points or stuff like that. I would not call it robust enough for `lasR` without further development.

Normalization of intensity using a range based correction is stable and could be implemented but without sensor tracking, it serves no purpose.

A stage to normalize intensity in 0-1 is also too much R&D for `lasR` in my opinion. First, intensity is an integer so [0,1] is not possible but we could normalize into [0, 2^16]. The _per file_ suggestion is a bad idea to me. You will end with variations between files that are worst that the original data. It should be using the global min max. Also `intensity` has usually a lot of outlier so min/max is not a good idea. Percentile is much better. So it requires a pre-read stage to get the global percentiles.

lidR is typically designed for this kind of usage while lasR is typically not designed for that. I already considered your suggestion on my side and for the moment `lasR` is still maturing  and intensity normalization is far in my priority list.

Now, as I previously said, if you want to sponsor the development, priorities may change. If you can support the software I can implement specific tools for you. This is part of my business plan for the future.
